### ISSUE TITLE
from Equilibration NPT to production NVT

### ISSUE THREAD

**Issue by anicoli** [openmm/openmm#3854]
Dear openMM developers,

I am writing my first scripts to run MD simulation with openMM, also following other examples from the community. Basically, we run an equilibration in the NVT ensemble (MonteCarloMembraneBarostat and LangevinIntegrator). Then, we saved the final state, checkpoint, and last frame in pdb format (that it is not wrapped).

```
simulation.reporters.append(app.CheckpointReporter(''+eq_folder+'/checkpnt.chk', trajectory_freq))
[...]
simulation.step(t4_free)   #simulatet4_free steps with no restraints

# Log file to check complete of simulation
with open(''+eq_folder+'/end_log.txt', 'w') as slog:
    slog.write('Simulation complete!')

#### Saving final files ####
# Writing last frame information of stride
final_state = simulation.context.getState(getPositions=True, getVelocities=True)
with open(final_state_file, 'w') as f:
    f.write(XmlSerializer.serialize(final_state))

last_frame = int(nsteps/trajectory_freq)
positions = simulation.context.getState(getPositions=True).getPositions()
PDBFile.writeFile(simulation.topology, positions, open(pdb_final, 'w'))
```
Now, we would like to start an MD production in the NVT ensemble (LangevinIntegrator) of the equilibrated system, but starting from the final state of the equilibration: last atomic positions, last velocities and box dimension. Here below, just an example that we started to write also following (#3782). Topology, initial pdb, and parameters are the same as the equilibration, and just need to re-import them in the new script for production. Final velocities and position, I can get them from the line simulation.loadState(final_state_file) #position and velocities to simulation.context.setVelocities(velocities), is this correct?  

```
# Load input file for MD production

input_coordinates='../CHARMM/input_file/test.pdb'
input_topology='../CHARMM/input_file/test.psf' 
input_par=[list_parameters]

# Check input files and stop in case of error
try:
    pdb = PDBFile(input_coordinates)
    psf = CharmmPsfFile(input_topology)
except:
    print("Problem loading files...")
    print("ending script...")

# Load final state 
simulation.loadState(final_state_file) #position and velocities
eq_state = simulation.context.getState(getVelocities=True, getPositions=True)
positions = eq_state.getPositions()
velocities = eq_state.getVelocities()
simulation.context.setPositions(positions)
simulation.context.setVelocities(velocities)

[...]

simulation = Simulation(topology, system, integrator, platform, platformProperties)
simulation.step(nsteps) 
```

The final box dimension from the equilibration is already implicit after importing the new atomic coordinates, or shall I need first to set them as psf.setBox(x,y,z), and then create a new system object and then get position and velocities from the state? or it is better to get the atomic position from the pdb of the last frame? 

```
psf.setBox(x,y,z)
system = psf.createSystem(...)
simulation = Simulation(topology, system, integrator, platform, platformProperties)
simulation.context.setPositions(positions)
```
Sorry for the many, question, I hope I explained the problem. Thank you in advance.
Best,
Alessandro

**Comment by peastman** [comment]
The saved state includes box vectors.  They'll get restored automatically by the call to `loadState()`.

I don't quite understand your code:

```python
eq_state = simulation.context.getState(getVelocities=True, getPositions=True)
positions = eq_state.getPositions()
velocities = eq_state.getVelocities()
simulation.context.setPositions(positions)
simulation.context.setVelocities(velocities)
```

Those lines do nothing at all.  They query the positions and velocities from the simulation, and then immediately call `setPositions()` and `setVelocities()` to set the same positions and velocities it already has?  All you really need to do is this:

```python
simulation = Simulation(topology, system, integrator, platform, platformProperties)
simulation.loadState(final_state_file)
```

Create a Simulation and call `loadState()` to restore all state information.  It should then be ready to simulate.

**Comment by anicoli** [comment]
Thank you for your fast reply @peastman .  Basically, when my equilibration in the NPT ensemble is finished, I save as stated above, the final state. Then, in the new script for MD production in the NVT ensemble I do the following passages :

1. Import initial pdb, psf and parameters
2. set inital dimension box
3. system = psf.createSystem(...) 
4. set integrator(...)
5. set Simulation(topology, system, integrator, platform, platformProperties)
6. simulation.loadState(final_state_file)
```
simulation = Simulation(topology, system, integrator, platform, platformProperties)
simulation.loadState(final_state_file)
```
7. Start the simulation

I am asking if this passage and the logic behind are correct, since in the equilibration stage/script, we have created, first the system object and then add MonteCarloMembraneBarostat and set the Langevin thermostat/integrator; instead in the MD production (NVT ensemble) we do not have the MonteCarloMembraneBarostat. Loading directly the final state of the equilibration into the script, cannot arise problems? since the two simulations object has different system features (NPT vs NVT).
Basically what I am trying to obtain is two independent scripts for NPT equilibration and NVT production, where the second start from the last state of the equilibration. I hope now is more clear.

**Comment by peastman** [comment]
Correct.  The call to `loadState()` will load all information stored in the State and set it as the current information in the new Simulation.  That includes positions and velocities, since you specified them when you created the State, and also certain information that always gets included in States: time, step count, and periodic box vectors.

**Comment by peastman** [comment]
See [this FAQ entry](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#nan).  It discusses causes of that error and how to debug them.

If you still can't figure it out, please open a new issue.  It isn't related to this one.
